internal/cache/file/downloader/parallel_downloads_job.go

```diff
@@ -0,0 +1,182 @@
+		if closeErr != nil {
+			logger.Errorf("Job:%p (%s:/%s) error while closing reader: %v", job, job.bucket.Name(), job.object.Name, closeErr)
+		}
+	}()
+
+	monitor.CaptureGCSReadMetrics(ctx, util.Parallel, end-start)
+
+	// Copy the contents from NewReader to cache file at appropriate offset.
+	offsetWriter := io.NewOffsetWriter(cacheFile, start)
+	_, err = io.CopyN(offsetWriter, newReader, end-start)
+	if err != nil {
+		err = fmt.Errorf("downloadRangeToFile: error at the time of copying content to cache file %w", err)
+	}
+	return err
+}
+
+// parallelDownloadObjectAsync does parallel download of the backing GCS object
+// into a file as part of file cache using multiple NewReader method of
+// gcs.Bucket running in parallel
+//
+// Note: There can only be one async parallel or non-parallel download running
+// for a job at a time.
+// Acquires and releases LOCK(job.mu)
+func (job *Job) parallelDownloadObjectAsync() {
+	// Cleanup the async job in all cases - completion/failure/invalidation.
+	defer job.cleanUpDownloadAsyncJob()
+
+	// Create, and open cache file for writing object into it.
+	cacheFile, err := cacheutil.CreateFile(job.fileSpec, os.O_TRUNC|os.O_RDWR)
+	if err != nil {
+		err = fmt.Errorf("parallelDownloadObjectAsync: error in creating cache file: %w", err)
+		job.failWhileDownloading(err)
+		return
+	}
+	defer func() {
+		err = cacheFile.Close()
+		if err != nil {
+			err = fmt.Errorf("parallelDownloadObjectAsync: error while closing cache file: %w", err)
+			job.failWhileDownloading(err)
+		}
+	}()
+
+	notifyInvalid := func() {
+		job.mu.Lock()
+		job.status.Name = Invalid
+		job.notifySubscribers()
+		job.mu.Unlock()
+	}
+
+	var start, end int64
+	end = int64(job.object.Size)
+	var parallelReadRequestSize = int64(job.fileCacheConfig.ReadRequestSizeMB) * cacheutil.MiB
+	var maxSingleStepReadSize = parallelReadRequestSize * int64(job.fileCacheConfig.DownloadParallelismPerFile)
+
+	for {
+		select {
+		case <-job.cancelCtx.Done():
+			return
+		default:
+			if start < end {
+				// Parallely download different ranges not more than maxSingleStepReadSize
+				// and not using go routines more than job.downloadParallelism
+				downloadErrGroup, downloadErrGroupCtx := errgroup.WithContext(job.cancelCtx)
+
+				var singleStepReadSize int64 = 0
+				for goRoutineIdx := 0; (goRoutineIdx < job.fileCacheConfig.DownloadParallelismPerFile) &&
+					(singleStepReadSize < maxSingleStepReadSize) && (start < end); goRoutineIdx++ {
+					rangeStart := start
+					rangeEnd := min(rangeStart+parallelReadRequestSize, end)
+
+					downloadErrGroup.Go(func() error {
+						return job.downloadRangeToFile(downloadErrGroupCtx, cacheFile, rangeStart, rangeEnd)
+					})
+
+					singleStepReadSize = singleStepReadSize + rangeEnd - rangeStart
+					start = rangeEnd
+				}
+				// If any of the go routine failed, consider the async job failed.
+				err = downloadErrGroup.Wait()
+				if err != nil {
+					if errors.Is(err, context.Canceled) {
```

Comment:
why is context canceled means invalid job?

Reply:
the context passed is the context associated with the job and in the current job flows, context is canceled only when the job is Invalidated (during eviction).